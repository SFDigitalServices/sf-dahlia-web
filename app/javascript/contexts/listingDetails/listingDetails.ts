interface ListingDetails {
  /**  */
  uid: string

  /**  */
  email: string

  /**  */
  created_at: Date

  /**  */
  updated_at: Date
}

interface ListingDetailsData {
  data: any
  success: boolean
}

export { ListingDetails, ListingDetailsData }
