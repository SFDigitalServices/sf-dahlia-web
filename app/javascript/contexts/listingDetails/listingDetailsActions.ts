import { createAction } from "typesafe-actions"
import { RailsUnit } from "../../api/types/rails/listings/RailsUnit"
import { RailsAmiChart } from "../../api/types/rails/listings/RailsAmiChart"

export enum ListingDetailsActions {
  StartFetchingUnits = "SetFetchingUnits",
  FinishFetchingUnits = "FinishFetchingUnits",
  SetUnits = "SetUnits",
  SetAmiCharts = "SetAmiCharts",
  StartFetchingAmiCharts = "SetFetchingAmiCharts",
  SetFetchedAmiCharts = "SetFetchedAmiCharts",
  SetFetchingAmiChartsError = "SetFetchingAmiChartsError",
  SetFetchingUnitsError = "SetFetchingUnitsError",
  FinishFetchingAmiCharts = "FinishFetchingAmiCharts",
}

export const startFetchingUnits = createAction(ListingDetailsActions.StartFetchingUnits)()
export const setUnits = createAction(ListingDetailsActions.SetUnits)()
export const setAmiCharts = createAction(ListingDetailsActions.SetAmiCharts)()
export const finishFetchingAmiCharts = createAction(ListingDetailsActions.FinishFetchingAmiCharts)<
  RailsAmiChart[]
>()
export const setFetchedAmiCharts = createAction(ListingDetailsActions.SetFetchedAmiCharts)()
export const startFetchingAmiCharts = createAction(ListingDetailsActions.StartFetchingAmiCharts)()
export const finishFetchingUnits = createAction(ListingDetailsActions.FinishFetchingUnits)<
  RailsUnit[]
>()
export const setFetchingUnitsError = createAction(
  ListingDetailsActions.SetFetchingUnitsError
)<Error>()
export const setFetchingAmiChartsError = createAction(
  ListingDetailsActions.SetFetchingAmiChartsError
)<Error>()
