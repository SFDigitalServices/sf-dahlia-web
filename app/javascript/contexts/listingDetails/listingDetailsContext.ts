import { createContext } from "react"
import RailsUnit from "../../api/types/rails/listings/RailsUnit"
import { RailsAmiChart, RailsAmiChartMetaData } from "../../api/types/rails/listings/RailsAmiChart"

export type ContextProps = {
  units: RailsUnit[]
  amiCharts: RailsAmiChart[]
  fetchedAmiCharts: boolean
  fetchingAmiCharts: boolean
  fetchedUnits: boolean
  fetchingUnits: boolean
  fetchUnits: (listingId: string) => void
  fetchAmiCharts: (chartsToFetch: RailsAmiChartMetaData[]) => void
  fetchingAmiChartsError: Error
  fetchingUnitsError: Error
}

export default createContext<Partial<ContextProps>>({})
