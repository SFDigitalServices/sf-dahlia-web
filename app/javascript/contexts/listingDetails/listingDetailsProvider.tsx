import React, { useReducer } from "react"
import ListingDetailsReducer from "./listingDetailsReducer"
import ListingDetailsContext, { ContextProps } from "./listingDetailsContext"
import { getUnits, getAmiCharts } from "../../api/listingApiService"
import { RailsUnit } from "../../api/types/rails/listings/RailsUnit"
import { RailsAmiChart, RailsAmiChartMetaData } from "../../api/types/rails/listings/RailsAmiChart"
import {
  finishFetchingUnits,
  finishFetchingAmiCharts,
  startFetchingAmiCharts,
  startFetchingUnits,
  setFetchingAmiChartsError,
  setFetchingUnitsError,
} from "./listingDetailsActions"

interface ListingDetailsProviderProps {
  children?: React.ReactNode
}

const ListingDetailsProvider = (props: ListingDetailsProviderProps) => {
  const [state, dispatch] = useReducer(ListingDetailsReducer, {
    units: [],
    amiCharts: [],
    fetchingUnits: false,
    fetchedUnits: false,
    fetchingAmiCharts: false,
    fetchedAmiCharts: false,
    fetchingAmiChartsError: null,
    fetchingUnitsError: null,
  })

  const contextValues: ContextProps = {
    units: state.units,
    amiCharts: state.amiCharts,
    fetchedAmiCharts: state.fetchedAmiCharts,
    fetchingAmiCharts: state.fetchingAmiCharts,
    fetchedUnits: state.fetchedUnits,
    fetchingUnits: state.fetchingUnits,
    fetchingAmiChartsError: state.fetchingAmiChartsError,
    fetchingUnitsError: state.fetchingUnitsError,
    fetchUnits: (listingId: string) => {
      dispatch(startFetchingUnits())
      getUnits(listingId)
        .then((units: RailsUnit[]) => {
          dispatch(finishFetchingUnits(units))
        })
        .catch((error) => {
          console.error(error)
          dispatch(setFetchingUnitsError(error))
        })
    },
    fetchAmiCharts: (chartsToFetch: RailsAmiChartMetaData[]) => {
      dispatch(startFetchingAmiCharts())
      getAmiCharts(chartsToFetch)
        .then((amiCharts: RailsAmiChart[]) => {
          dispatch(finishFetchingAmiCharts(amiCharts))
        })
        .catch((error) => {
          console.error(error)
          dispatch(setFetchingAmiChartsError(error))
        })
    },
  }

  return (
    <ListingDetailsContext.Provider value={contextValues}>
      {props.children}
    </ListingDetailsContext.Provider>
  )
}

export default ListingDetailsProvider
