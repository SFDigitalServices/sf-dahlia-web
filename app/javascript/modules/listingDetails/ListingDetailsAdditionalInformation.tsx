import React from "react"
import { LinkButton, ListingDetailItem, t } from "@bloom-housing/ui-components"
import { RailsListing } from "../listings/SharedHelpers"
import { TextTruncate } from "../../components/TextTruncate"
import { isHabitatListing, isSale } from "../../util/listingUtil"
import { stripMostTags } from "../../util/filterUtil"
import { getTranslatedString } from "../../util/languageUtil"

export interface ListingDetailsAdditionalInformationProps {
  listing: RailsListing
  imageSrc: string
}

export const ListingDetailsAdditionalInformation = ({
  listing,
  imageSrc,
}: ListingDetailsAdditionalInformationProps) => {
  const getCommissionString = () => {
    return listing.Realtor_Commission_Unit === "percent"
      ? t("listings.realtorCommissionPercentage", {
          percentage: listing.Realtor_Commission_Amount,
        })
      : `$${listing.Realtor_Commission_Amount.toLocaleString()}`
  }

  return (
    <ListingDetailItem
      imageAlt={""}
      imageSrc={imageSrc}
      title={t("listings.additionalInformation.header")}
      subtitle={t("listings.additionalInformation.subheader")}
    >
      <div className="listing-detail-panel">
        {listing.Listing_Other_Notes && (
          <div className="info-card bg-gray-100 border-0">
            <h3 className="text-serif-xl">{t("listings.specialNotes")}</h3>
            <TextTruncate
              className="primary-lighter-markup-link translate"
              buttonClassName="text-blue-700"
              text={stripMostTags(
                getTranslatedString(
                  listing.Listing_Other_Notes,
                  "Listing_Other_Notes",
                  listing.translations
                )
              )}
            />
          </div>
        )}
        {(!!listing.Required_Documents || isSale(listing)) && (
          <div className="info-card bg-gray-100 border-0">
            <h3 className="text-serif-xl">{t("listings.requiredDocuments")}</h3>
            <div className="text-xs">
              <TextTruncate
                className="primary-lighter-markup-link translate"
                buttonClassName="text-blue-700"
                text={stripMostTags(
                  getTranslatedString(
                    listing.Required_Documents,
                    "Required_Documents",
                    listing.translations
                  )
                )}
              />
            </div>
            {isSale(listing) && !isHabitatListing(listing) && (
              <div className="text-xs mt-4">
                <TextTruncate
                  className="primary-lighter-markup-link translate"
                  buttonClassName="text-blue-700"
                  text={t("listings.requiredDocumentsAfterApplying", {
                    url: "https://sfmohcd.org/after-homebuyer-lottery",
                  })}
                />
              </div>
            )}
          </div>
        )}
        {listing.Legal_Disclaimers && (
          <div className="info-card bg-gray-100 border-0">
            <h3 className="text-serif-xl">{t("listings.importantProgramRules")}</h3>
            <div className="text-xs">
              <TextTruncate
                text={stripMostTags(
                  getTranslatedString(
                    listing.Legal_Disclaimers,
                    "Legal_Disclaimers",
                    listing.translations
                  )
                )}
                className="primary-lighter-markup-link translate"
                buttonClassName="text-blue-500"
              />
            </div>
          </div>
        )}
        {listing.CC_and_R_URL && (
          <div className="info-card bg-gray-100 border-0">
            <h3 className="text-serif-xl">{t("listings.cc&r")}</h3>
            <div className="text-xs">
              <TextTruncate
                className="primary-lighter-markup-link translate"
                buttonClassName="text-blue-500"
                text={t("listings.cc&rDescription")}
              />
              <br />
              <LinkButton href={listing.CC_and_R_URL} className={"mt-4"} newTab={true}>
                {t("listings.downloadPdf")}
              </LinkButton>
            </div>
          </div>
        )}
        {isSale(listing) && (
          <div className="info-card bg-gray-100 border-0">
            <h3 className="text-serif-xl">{t("listings.realtorCommission")}</h3>
            <div className="text-xs">
              {listing.Allows_Realtor_Commission ? (
                <>
                  <span className={"font-bold"}>{t("listings.realtorCommissionHeader")}</span>
                  {getCommissionString()}
                </>
              ) : (
                t("listings.realtorCommissionNotEligible")
              )}
              {listing.Realtor_Commission_Info && (
                <div className={"mt-4"}>
                  <span className={"font-bold"}>{t("listings.realtorCommissionHowTo")}</span>
                  <span>
                    <TextTruncate
                      className="primary-lighter-markup-link translate"
                      buttonClassName="text-blue-500"
                      text={stripMostTags(
                        getTranslatedString(
                          listing.Realtor_Commission_Info,
                          "Realtor_Commission_Info",
                          listing.translations
                        )
                      )}
                    />
                  </span>
                </div>
              )}
            </div>
          </div>
        )}
        {/* TODO: implement once we've established needs for sales listings
              {listing.isSale && (
                <div className="info-card bg-gray-100 border-0">
                  <h3 className="text-serif-xl">For the Buyer's Realtor</h3>
                  {listing.Allows_Realtor_Commission ? (
                    display realtor_commission_header
                    realtorComissionMessage
                    {listing.Realtor_Commission_Info && realtor_commission_how_to}
                  ) : display realtor_commission_not_eligible message}
                </div>
              )}
            */}
        {listing.Repricing_Mechanism && (
          <div className="info-card bg-gray-100 border-0">
            <h3 className="text-serif-xl">{t("listings.rePricing")}</h3>
            <div className="text-xs">
              <TextTruncate
                className="primary-lighter-markup-link translate"
                buttonClassName="text-blue-500"
                text={stripMostTags(
                  getTranslatedString(
                    listing.Repricing_Mechanism,
                    "Repricing_Mechanism",
                    listing.translations
                  )
                )}
              />
            </div>
          </div>
        )}
      </div>
    </ListingDetailItem>
  )
}
