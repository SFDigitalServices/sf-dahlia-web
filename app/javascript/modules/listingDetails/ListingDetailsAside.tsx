import React from "react"
import { RailsListing } from "../listings/SharedHelpers"
import { ListingDetailsInfoSession } from "../listingDetailsAside/ListingDetailsInfoSession"
import { ListingDetailsApply } from "../listingDetailsAside/ListingDetailsApply"
import { ListingDetailsProcess } from "../listingDetailsAside/ListingDetailsProcess"
import { isOpen } from "../../util/listingUtil"
import { ListingDetailsApplicationDate } from "../listingDetailsAside/ListingDetailsApplicationDate"

export interface ListingDetailsSidebarProps {
  listing: RailsListing
}

export const ListingDetailsAside = ({ listing }: ListingDetailsSidebarProps) => {
  const isApplicationOpen = isOpen(listing)

  return (
    <aside className="w-full static md:absolute md:right-0 md:w-1/3 md:top-0 sm:w-2/3 md:ml-2 h-full md:border border-solid bg-white">
      <div className="hidden md:block">
        <ListingDetailsApplicationDate listing={listing} />
        {isApplicationOpen && <ListingDetailsInfoSession listing={listing} />}
        <ListingDetailsApply listing={listing} />
        <ListingDetailsProcess listing={listing} />
      </div>
    </aside>
  )
}
