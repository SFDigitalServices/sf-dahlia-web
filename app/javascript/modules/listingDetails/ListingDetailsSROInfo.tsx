import React from "react"
import { InfoCard, t } from "@bloom-housing/ui-components"
import { RailsListing } from "../listings/SharedHelpers"
import { isPluralSRO } from "../../util/listingUtil"

export interface ListingDetailsSROInfo {
  listing: RailsListing
}

export const ListingDetailsSROInfo = ({ listing }: ListingDetailsSROInfo) => (
  <InfoCard title={t("listings.singleRoomOccupancy")}>
    {!isPluralSRO(listing)
      ? t("listings.singleRoomOccupancyDescription")
      : t("listings.merryGoRoundSingleRoomOccupancyDescription")}
  </InfoCard>
)
