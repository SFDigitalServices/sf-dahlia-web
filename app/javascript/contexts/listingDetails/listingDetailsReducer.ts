import { createReducer } from "typesafe-actions"
import { ListingDetailsActions } from "./listingDetailsActions"
import { RailsUnit } from "../../api/types/rails/listings/RailsUnit"
import { RailsAmiChart } from "../../api/types/rails/listings/RailsAmiChart"
import { AmiChartData } from "./listingDetailsContext"

type ListingDetailsState = {
  fetchingUnits: boolean
  fetchedUnits: boolean
  fetchingAmiCharts: boolean
  fetchedAmiCharts: boolean
  units: RailsUnit[]
  amiChartData: AmiChartData
  amiCharts: RailsAmiChart[]
  fetchingAmiChartsError: Error
  fetchingUnitsError: Error
}

const ListingDetailsReducer = createReducer({ units: [] } as ListingDetailsState, {
  [ListingDetailsActions.StartFetchingUnits]: (state) => {
    return { ...state, fetchingUnits: true }
  },
  [ListingDetailsActions.StartFetchingAmiCharts]: (state) => {
    return { ...state, fetchingAmiCharts: true }
  },
  [ListingDetailsActions.FinishFetchingUnits]: (state, { payload: units }) => {
    const amiChartYears: Array<number> = []
    const amiChartTypes = []
    const amiChartPercentages = []

    units.forEach((unit: RailsUnit) => {
      const amiChartYear = amiChartYears.find((year) => {
        return year === unit.AMI_chart_year
      })

      const amiChartType = amiChartTypes.find((type) => {
        return type === unit.AMI_chart_type
      })

      const amiChartPercentage = amiChartPercentages.find((type) => {
        return type === unit.Max_AMI_for_Qualifying_Unit
      })

      if (!amiChartYear) {
        amiChartYears.push(unit.AMI_chart_year)
      }

      if (!amiChartType) {
        amiChartTypes.push(unit.AMI_chart_type)
      }

      if (!amiChartPercentage) {
        amiChartPercentages.push(unit.Max_AMI_for_Qualifying_Unit)
      }
    })

    return {
      ...state,
      fetchingUnits: false,
      fetchedUnits: true,
      units,
      amiChartData: {
        years: amiChartYears,
        percentages: amiChartPercentages,
        types: amiChartTypes,
      },
    }
  },
  [ListingDetailsActions.FinishFetchingAmiCharts]: (state, { payload: amiCharts }) => {
    return {
      ...state,
      fetchingAmiCharts: false,
      fetchedAmiCharts: true,
      amiCharts: amiCharts?.map((amiChart: RailsAmiChart) => {
        return {
          ...amiChart,
          chartType: amiChart?.values[0]?.chartType,
        }
      }),
    }
  },
})

export default ListingDetailsReducer
