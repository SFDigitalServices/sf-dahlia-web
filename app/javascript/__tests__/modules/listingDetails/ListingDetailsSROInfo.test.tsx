import React from "react"
import { render } from "@testing-library/react"
import { ListingDetailsSROInfo } from "../../../modules/listingDetails/ListingDetailsSROInfo"
import {
  sroRentalListing,
  pluralSroRentalListing,
} from "../../data/RailsRentalListing/listing-rental-sro"

describe("ListingDetailsLotteryInfo", () => {
  const OLD_ENV = process.env

  beforeEach(() => {
    jest.resetModules()
    process.env = { ...OLD_ENV }
  })

  afterAll(() => {
    process.env = OLD_ENV
  })

  it("displays the SRO info box for a regular SRO", () => {
    // This component pulls in react-media, which needs this custom mock
    window.matchMedia = jest.fn().mockImplementation((query) => {
      return {
        matches: true,
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      }
    })

    const { asFragment } = render(<ListingDetailsSROInfo listing={sroRentalListing} />)

    expect(asFragment()).toMatchSnapshot()
  })

  it("displays the Merry Go Round Housing description for its listing", () => {
    const listing = { ...pluralSroRentalListing, Id: "a0W0P00000F7t4uUAB" }
    // This component pulls in react-media, which needs this custom mock
    window.matchMedia = jest.fn().mockImplementation((query) => {
      return {
        matches: true,
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      }
    })

    const { asFragment } = render(<ListingDetailsSROInfo listing={listing} />)

    expect(asFragment()).toMatchSnapshot()
  })
})
