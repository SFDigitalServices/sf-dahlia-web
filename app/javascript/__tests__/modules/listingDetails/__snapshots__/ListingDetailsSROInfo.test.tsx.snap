// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingDetailsLotteryInfo displays the Merry Go Round Housing description for its listing 1`] = `
<DocumentFragment>
  <div
    class="info-card"
  >
    <div
      class="info-card__header"
    >
      <h3
        class="info-card__title"
      >
        Single Room Occupancy
      </h3>
    </div>
    <div
      class="markdown info-card__content"
    >
      <span>
        This property offers single rooms for up to 2 people only. Tenants may share bathrooms, and sometimes kitchen facilities.
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ListingDetailsLotteryInfo displays the SRO info box for a regular SRO 1`] = `
<DocumentFragment>
  <div
    class="info-card"
  >
    <div
      class="info-card__header"
    >
      <h3
        class="info-card__title"
      >
        Single Room Occupancy
      </h3>
    </div>
    <div
      class="markdown info-card__content"
    >
      <span>
        This property offers single rooms for one person only. Tenants may share bathrooms, and sometimes kitchen facilities.
      </span>
    </div>
  </div>
</DocumentFragment>
`;
